Npgsql/NpgsqlLargeObjectStream.cs

```diff
@@ -0,0 +1,306 @@
+            }
+        }
+
+        /// <summary>
+        /// CanTimeout always returns false.
+        /// </summary>
+        public override bool CanTimeout
+        {
+            get
+            {
+                return false; // TODO
+            }
+        }
+
+        /// <summary>
+        /// CanRead always returns true, unless the stream has been closed.
+        /// </summary>
+        public override bool CanRead
+        {
+            get { return true && !_disposed; }
+        }
+
+        /// <summary>
+        /// CanWrite returns true if the stream was opened with write permissions, and the stream has not been closed.
+        /// </summary>
+        public override bool CanWrite
+        {
+            get { return _writeable && !_disposed; }
+        }
+
+        /// <summary>
+        /// CanSeek always returns true, unless the stream has been closed.
+        /// </summary>
+        public override bool CanSeek
+        {
+            get { return true && !_disposed; }
+        }
+
+        /// <summary>
+        /// Returns the current position in the stream. Getting the current position does not need a round-trip to the server, however setting the current position does.
+        /// </summary>
+        public override long Position
+        {
+            get
+            {
+                CheckDisposed();
+                return _pos;
+            }
+            set
+            {
+                Seek(value, SeekOrigin.Begin);
+            }
+        }
+
+        /// <summary>
+        /// Gets the length of the large object. This internally seeks to the end of the stream to retrieve the length, and then back again.
+        /// </summary>
+        public override long Length
+        {
+            get { return GetLengthInternal(); }
+        }
+
+        // TODO: uncomment this
+        /*public Task<long> GetLengthAsync()
+        {
+            return GetLengthInternalAsync();
+        }*/
+
+        [GenerateAsync]
+        long GetLengthInternal()
+        {
+            CheckDisposed();
+            long old = _pos;
+            long retval = Seek(0, SeekOrigin.End);
+            if (retval != old)
+                Seek(old, SeekOrigin.Begin);
+            return retval;
+        }
+
+        /// <summary>
+        /// Seeks in the stream to the specified position. This requires a round-trip to the backend.
+        /// </summary>
+        /// <param name="offset">A byte offset relative to the <i>origin</i> parameter.</param>
+        /// <param name="origin">A value of type SeekOrigin indicating the reference point used to obtain the new position.</param>
+        /// <returns></returns>
+        [GenerateAsync]
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            if (origin < SeekOrigin.Begin || origin > SeekOrigin.End)
+                throw new ArgumentException("Invalid origin");
+            if (!_manager.Has64BitSupport && offset != (long)(int)offset)
```

Comment:
It's actually a good check to have inside the contract block. You just have to check a public Has64BitSupport property directly on the stream rather than _manager (and that property can internally go to _manager or anything else it wants).